packages/@aws-c2a/cdk-pipelines-step/lib/private/change-analysis-check.ts

```diff
@@ -0,0 +1,177 @@
+import * as s3 from '@aws-cdk/aws-s3';
+import { RemovalPolicy, Tags } from '@aws-cdk/core';
+import { Construct } from 'constructs';
+import { PreApproveLambda } from './pre-approve-lambda';
+import { ifElse } from './security-check-utils';
+
+// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
+// eslint-disable-next-line no-duplicate-imports, import/order
+import { Construct as CoreConstruct } from '@aws-cdk/core';
+
+/**
+ * Properties for an ChangeAnalysisCheck
+ */
+export interface ChangeAnalysisCheckProps {
+  /**
+   * The pipeline that will be automatically approved
+   *
+   * Will have a tag added to it.
+   */
+  readonly codePipeline: cp.Pipeline;
+}
+
+/**
+ * A construct containing both the Lambda and CodeBuild Project
+ * needed to conduct a security check on any given application stage.
+ *
+ * The Lambda acts as an auto approving mechanism that should only be
+ * triggered when the CodeBuild Project registers no security changes.
+ *
+ * The CodeBuild Project runs c2a diff on the application stage,
+ * and exports the link to the console of the project.
+ */
+export class ChangeAnalysisCheck extends CoreConstruct {
+  /**
+   * A lambda function that approves a Manual Approval Action, given
+   * the following payload:
+   *
+   * {
+   *  "PipelineName": [CodePipelineName],
+   *  "StageName": [CodePipelineStageName],
+   *  "ActionName": [ManualApprovalActionName]
+   * }
+   */
+  public readonly preApproveLambda: lambda.Function;
+  /**
+   * A CodeBuild Project that runs c2a diff on the application stage.
+   *
+   * - If the c2a diff registers no security changes, CodeBuild will invoke the
+   *   pre-approval lambda and approve the ManualApprovalAction.
+   * - If changes are detected, CodeBuild will exit into a ManualApprovalAction
+   */
+  public readonly c2aDiffProject: codebuild.Project;
+  /**
+   * A S3 Bucket that stores all the generated web apps created from
+   * the change reports.
+   */
+  public readonly webappBucket: s3.Bucket;
+
+  constructor(scope: Construct, id: string, props: ChangeAnalysisCheckProps) {
+    super(scope, id);
+
+    Tags.of(props.codePipeline).add('CHANGE_ANALYSIS', 'ALLOW_APPROVE', {
+      includeResourceTypes: ['AWS::CodePipeline::Pipeline'],
+    });
+
+    const { preApproveLambda, invokeLambda } = new PreApproveLambda(this, 'C2APreApproveLambda', {
+      pipelineTag: 'CHANGE_ANALYSIS',
+    });
+    this.preApproveLambda = preApproveLambda;
+
+    this.webappBucket = new s3.Bucket(this, 'C2AWebappBucket', {
+      publicReadAccess: false,
+      autoDeleteObjects: true,
+      removalPolicy: RemovalPolicy.DESTROY,
+    });
+
+    const putObject =
+      'aws s3api put-object' +
+      ` --bucket ${this.webappBucket.bucketName}` +
+      ' --key $STAGE_NAME/index.html' +
+      ' --body index.html' +
+      ' --content-type text/html';
+
+    const signObject =
+      'aws s3 presign' +
```

Comment:
Almost there! But slight complication: if you sign with the *current* credentials (which are short-lived), the link will never be able to live for longer than 1hr. Even `--expires-in` doesn't help, it cannot extend the lifetime of the signed link for longer than the lifetime of the credentials you used to sign the link.

We will need to create an **IAM User** as well (with permissions to read the S3 bucket), generate a public/private key pair for it, and use those credentials to sign the link. See the link I sent you privately. It's worth extracting that into its own construct as well, for neatness' sake (otherwise it's going to become messy in here).